caddyhttp/httpserver/plugin.go

```diff
@@ -51,7 +51,8 @@ var contexts []*httpContext
 
 func newContext() caddy.Context {
 	context := &httpContext{keysToSiteConfigs: make(map[string]*SiteConfig)}
-	contexts = append(contexts, context)
+	// put the new context at start to allow setup of directives on new instance
+	contexts = append([]*httpContext{context}, contexts...)
```

Comment:
In case caddy.Start() is called more than once; it would definitely be unusual but since Caddy can be used as a library I'm not gonna rule it out.